### ISSUE TITLE
Can't sign into account in tor browser

### ISSUE THREAD

**Issue by fifn2** [tootsuite/mastodon#14910]
In tor browser, I sign in to mastodon.social, but I get the error:
```
Due to a bug in our code or a browser compatibility issue, this page could not be displayed correctly.

Try refreshing the page. If that does not help, you may still be able to use Mastodon through a different browser or native app.
``` 
The error stack information is:
```
TypeError: e is undefined


nn@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:950305
Yi@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1002105
lu@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022535
cu@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022424
su@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022277
Zs@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1019307
Va/<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970949
t.unstable_runWithPriority@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1046192
Ba@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970658
Va@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970896
Ka@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970829
eu@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1019598
notify@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100291
c</t.notifyNestedSubs@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100926
c</t.handleChangeWrapper@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100994
g@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:385885
pt/</</<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:358186
pt/</</<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:357730
pt/</</</<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:357483
a/</</<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:315329
dispatch@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:388056
s/</<@https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:388873



nn (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:950305)
Yi (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1002105)
lu (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022535)
cu (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022424)
su (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1022277)
Zs (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1019307)
Va/< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970949)
t.unstable_runWithPriority (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1046192)
Ba (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970658)
Va (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970896)
Ka (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:970829)
eu (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:1019598)
notify (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100291)
c</t.notifyNestedSubs (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100926)
c</t.handleChangeWrapper (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:100994)
g (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:385885)
pt/</</< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:358186)
pt/</</< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:357730)
pt/</</</< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:357483)
a/</</< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:315329)
dispatch (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:388056)
s/</< (https://static-cdn.mastodon.social/packs/js/common-48cf90a88c7bc543238f.js:2:388873)
```
What's going on?

**Comment by ThibG** [comment]
It's hard to say… for some reason the traceback couldn't be properly mapped and I'm unable to make sense of it… so many things could be happening… did you successfully use your account with the Tor browser before or did it break recently? Which version of the Tor browser are you using? Do you also log in with another browser? If so, does it work? Are you using the advanced view or the single-column view? In case it's the advanced view, which columns are set up?

**Comment by fifn2** [comment]
> Did you successfully use your account with the Tor browser before or did it break recently? 
I have never successfully logged into mastodon.social on Tor browser.
>  Which version of the Tor browser are you using?
78.3.0esr
> Do you also log in with another browser? If so, does it work?
Yes, I can log in just fine on stock Firefox
> Are you using the advanced view or the single-column view? In case it's the advanced view, which columns are set up?
I'm using the default view

**Comment by Gargron** [comment]
I think the ESR part is the problem, we've already had problems with our JS not working in ESR version of Firefox.

**Comment by fifn2** [comment]
Ah. I'll switch from Debian soon so I can test out a newer version of Firefox.

**Comment by ThibG** [comment]
78.3.0esr should work, and we should make sure ESR versions work anyway.
I can't reproduce this on my instance, not sure exactly which version mastodon.social is currently running.

**Comment by tech-exorcist** [comment]
I have the same problem on mastodon.uno, with Tor Browser 10.0.2 (GNU/Linux version), `Safest` security mode.
NoScript is at version 11.1.3, and I tried both setting mastodon.uno to "Trusted" and using "Disable restrictions for this tab".
The error page on mastodon.uno says that the server is running Mastodon v3.2.1.
The error message is the same that @fifn2 pasted (but in Italian, since mastodon.uno is in Italian).
Here's my stacktrace:

```
TypeError: e is undefined


nn@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:950777
Yi@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1002577
lu@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1023007
cu@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1022896
su@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1022749
Zs@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1019779
Va/<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:971421
t.unstable_runWithPriority@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1046664
Ha@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:971130
Va@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:971368
Ka@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:971301
eu@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:1020070
notify@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:100257
c</t.notifyNestedSubs@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:100892
c</t.handleChangeWrapper@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:100960
g@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:377494
pt/</</<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:359873
pt/</</<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:359417
pt/</</</<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:359170
a/</</<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:317016
dispatch@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:379665
s/</<@https://mastodon.uno/packs/js/common-fea69b14509bad28a4dd.js:2:380482



oc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:971:73)
Lg (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:4323:561)
Xi (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5561:4)
b.forEach(function (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5551:25)
Ui (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5545:2)
c.parentNode.insertBefore (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5352:400)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2254:6)
e (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:7162:8)
tc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2239:11)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2251:16)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2248:73)
Mi (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5366:15)
batch(function (webpack:///node_modules/react-redux/es/utils/Subscription.js:19:6)
notifyNestedSubs (webpack:///node_modules/react-redux/es/utils/Subscription.js:90:19)
handleChangeWrapper (webpack:///node_modules/react-redux/es/utils/Subscription.js:95:11)
Error (webpack:///node_modules/redux/es/redux.js:213:50)
audio.pause (webpack:///app/javascript/mastodon/middleware/sounds.js:16:6)
action.type.match (webpack:///app/javascript/mastodon/middleware/errors.js:10:17)
action.type.match (webpack:///app/javascript/mastodon/middleware/loading_bar.js:18:29)
function (webpack:///node_modules/redux-thunk/es/index.js:5:21)
_dispatch (webpack:///node_modules/redux/es/redux.js:634:24)
fetchTrends (webpack:///app/javascript/mastodon/actions/trends.js:11:5)
```

**Comment by kushaldas** [comment]
> I have the same problem on mastodon.uno, with Tor Browser 10.0.2 (GNU/Linux version), `Safest` security mode.

Javascript is disabled in the `safest` security mode in the Tor Browser. Any web application which is dependent on JS on the client will not behave well in `safest` mode.

**Comment by ThibG** [comment]
I'm afraid I can't reproduce this either and I can't manage to map the traceback to something that makes sense :/

**Comment by ThibG** [comment]
To anyone affected by this issue:
- does the issue occur immediately when signing in, or after some interaction? if so, what interaction?
- does it always occur, or only some of the time?
- are you using the “advanced web interface” (see `/settings/preferences/appearance`)?
- does it work without the Tor Browser?
- when the error occurs, can you click it in the developer tools (F12) and take a screenshot of the debugger?

**Comment by gan60kug** [comment]
Same for me:
Tor Browser 10.0.12
Security level: Safest
JavaScript enabled manually for the mastodon instance domain
Mastodon v3.3.0

---

> does the issue occur immediately when signing in, or after some interaction? if so, what interaction?
> does it always occur, or only some of the time?

When I try to sign in directly on the mastodon instance homepage - yes
When I try to sign in, while being on a profile page - no

Side effect: the warning appears: "A webpage is slowing down your browser".

After having logged in via second variant, navigating to the instance homepage again triggers this error. Also: When I intend to reply to some toot, starting on a profile page, the site crashes.

> are you using the “advanced web interface” (see /settings/preferences/appearance)?

No.

> does it work without the Tor Browser?

Probably. It also works on Security Level Safer.

> when the error occurs, can you click it in the developer tools (F12) and take a screenshot of the debugger?

The debugger lands in webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:
```
function oc(a) {
  if (!ya) return !1;
  a = "on" + a;
  var b = (a in document);
  b || (b = document.createElement("div"), b.setAttribute(a, "return;"), b = "function" === typeof b[a]);
  return b;
}
```

Error stack trace seems to be equivalent to the one from @tech-exorcist :

>oc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:971:61)
Lg (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:4323:558)
Xi (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5562:4)
b.forEach(function (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5551:39)
Ui (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5545:18)
c.parentNode.insertBefore (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5352:400)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2254:4)
e (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:7162:10)
tc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2237:55)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2251:8)
dc (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:2248:73)
Mi (webpack:///node_modules/react-dom/cjs/react-dom.production.min.js:5367:4)
batch(function (webpack:///node_modules/react-redux/es/utils/Subscription.js:19:6)
notifyNestedSubs (webpack:///node_modules/react-redux/es/utils/Subscription.js:90:19)
handleChangeWrapper (webpack:///node_modules/react-redux/es/utils/Subscription.js:95:11)
Error (webpack:///node_modules/redux/es/redux.js:213:36)
audio.pause (webpack:///app/javascript/mastodon/middleware/sounds.js:15:21)
action.type.match (webpack:///app/javascript/mastodon/middleware/errors.js:10:17)
action.type.match (webpack:///app/javascript/mastodon/middleware/loading_bar.js:18:29)
function (webpack:///node_modules/redux-thunk/es/index.js:5:11)
_dispatch (webpack:///node_modules/redux/es/redux.js:634:24)
fetchTrends (webpack:///app/javascript/mastodon/actions/trends.js:11:5)

**Comment by ClearlyClaire** [comment]
> Security level: Safest
> JavaScript enabled manually for the mastodon instance domain

How do you manually enable Javascript for a given domain in Safest security level?

**Comment by gan60kug** [comment]
@ClearlyClaire I just set the domain entry to Temp. Trusted - no check of 'Override of Tor Browser's security level preset'.

**Comment by ClearlyClaire** [comment]
Are you sure you're using Tor Browser 10.0.12? I can't find those settings anywhere at all.

**Comment by gan60kug** [comment]
@ClearlyClaire I mean following addon UI (taken from the official site): 
![](https://noscript.net/noscript/ss0.png)

Tor Browser by default does not show the NoScript menu icon. If not shown, press "Customize..." in somewhere in Menu bar (where all icons reside) and drag the NoScript icon from the popping up dialog to the menu. Then you will be able to manually enable JavaScript for the mastodon domain instance.

**Comment by ClearlyClaire** [comment]
@gan60kug thank you, I can now reproduce this. I will investigate, but this seems to be something weird within React itself rather than Mastodon.

**Comment by ClearlyClaire** [comment]
Ok, so what happens is that Tor Browser's “Safest” security mode disables SVGs entirely, causing various issues throughout the Mastodon WebUI. One of those issues is that when rendering the `Sparklines` component for hashtags, React attempts to create an `svg` element, which doesn't work, then attempts to set properties on it, which fails with an exception.

I'm inclined to just declare Mastodon's WebUI as not supporting Tor Browser's Safest mode, as SVGs are used throughout the interface. However, I had already made most of the work, so I submitted PR #15809 to simply catch and ignore such errors when rendering hashtag components. This will fix the crash, but you will still have some issues with missing icons and stuff.
